### PR TITLE
posix: implement chown, fchown, lchown

### PR DESCRIPTION
**Pull request by marcinslusarz**
Built on top of PR #50

---
This change is [Reviewable](https://reviewable.io/reviews/marcinslusarz/pmemfile/56)

### CODE REVIEW

File: src/libpmemfile-core/file.c [hunk 1]
```diff
@@ -1674,3 +1674,158 @@ pmemfile_truncate(PMEMfilepool *pfp, const char *path, off_t length)
...
+					!gid_in_list(cred, group))
+				pmemfile_tx_abort(EPERM);
+		}
+
+		COMPILE_ERROR_ON(offsetof(struct pmemfile_inode, gid) !=
+				offsetof(struct pmemfile_inode, uid) +
+				sizeof(inode->uid));
+
+		pmemobj_tx_add_range_direct(&inode->uid,
+				sizeof(inode->uid) + sizeof(inode->gid));
```
Comment (GBuella) [inline src/libpmemfile-core/file.c]: I don't think we need to care a lot about chown being 'fast', but whatever...

File: src/libpmemfile-core/file.c [hunk 2]
```diff
@@ -1674,3 +1674,158 @@ pmemfile_truncate(PMEMfilepool *pfp, const char *path, off_t length)
...
+vinode_chown(PMEMfilepool *pfp, struct pmemfile_cred *cred,
+		struct pmemfile_vinode *vinode, uid_t owner, gid_t group)
+{
+	struct pmemfile_inode *inode = vinode->inode;
+	int error = 0;
+
+	os_rwlock_wrlock(&vinode->rwlock);
+
+	TX_BEGIN_CB(pfp->pop, cb_queue, pfp) {
+		if (!(cred->caps & (1 << PMEMFILE_CAP_CHOWN))) {
```
Comment (GBuella) [inline src/libpmemfile-core/file.c]: These ten lines can be moved, to be outside the transaction. Also, probably there would be no need to split the 'if (owner != -1 && owner != cred->fsuid)' line into two lines, if it had one fewer tab in front of it.

File: src/libpmemfile-core/file.c [hunk 3]
```diff
@@ -1674,3 +1674,171 @@ pmemfile_truncate(PMEMfilepool *pfp, const char *path, off_t length)
...
+{
+	struct pmemfile_inode *inode = vinode->inode;
+	int error = 0;
+
+	if (owner == (uid_t)-1 && group == (gid_t)-1)
+		return 0;
+
+	os_rwlock_wrlock(&vinode->rwlock);
+
+	if (!(cred->caps & (1 << PMEMFILE_CAP_CHOWN))) {
```
Comment (GBuella) [inline src/libpmemfile-core/file.c]: Well, not that it matters a lot, but this section of code doesn't seem to use the vinode either...
So why lock the vinode before it?

Comment (GBuella) [inline src/libpmemfile-core/file.c]: All these 'error = EXX; goto end;' pairs can be replaced with 'return EXX' statements.
The 'end' label can be removed, and locking can be simply around the transaction.
Simpler code - better code, yupee

File: src/libpmemfile-core/file.c [hunk 4]
```diff
@@ -1674,3 +1674,171 @@ pmemfile_truncate(PMEMfilepool *pfp, const char *path, off_t length)
...
+
+		if (group != (gid_t)-1 && group != inode->gid) {
+			if (group != cred->fsgid && !gid_in_list(cred, group)) {
+				error = EPERM;
+				goto end;
+			}
+		}
+	}
+
+	COMPILE_ERROR_ON(offsetof(struct pmemfile_inode, gid) !=
```
Comment (GBuella) [inline src/libpmemfile-core/file.c]: But this assertion might actually better be right next to the pmemobj_tx_add_range_direct call, to make it even more obvious why it is needed. It doesn't do anything runtime anyways.

File: src/libpmemfile-core/file.c (hunk 3 above)
Comment (marcinslusarz) [inline src/libpmemfile-core/file.c]: Unfortunately this code uses inode->uid/gid, so I can't move it above the lock.

File: src/libpmemfile-core/file.c (hunk 4 above)
Comment (marcinslusarz) [inline src/libpmemfile-core/file.c]: Done.

File: src/libpmemfile-core/file.c (hunk 3 above)
Comment (GBuella) [inline src/libpmemfile-core/file.c]: ooooooh, I see now, sorry

**Pull request merged**